File: UnityHVOF/Assets/Scripts/Infrastructure/Solvers/FDMCoatingSolver2D.cs
// ============================================================================
// 2D Finite Difference Coating Solver — Stage 2 Laser Treatment.
//
// Specialized 2D FDM solver for laser treatment of HVOF coatings.
// Differs from FDMHeatSolver3D:
//   1. Operates on a 2D coating grid with spatially varying k (pore-aware)
//   2. Uses Beer-Lambert volumetric absorption: Q ~ exp(-mu*y)
//   3. Tracks cumulative melt history per cell
//   4. Handles coating + substrate as distinct material regions
//
// Governing equation:
//     rho*cp * dT/dt = d/dx(k*dT/dx) + d/dy(k*dT/dy) + Q(x,y,t)
//
// Port of: infrastructure/solvers/fdm_coating_solver_2d.py (222 lines)
// ============================================================================

using System;
using System.Diagnostics;
using HVOFSim.Domain.Entities;

namespace HVOFSim.Infrastructure.Solvers
{
    /// <summary>
    /// 2D transient heat solver for laser treatment of HVOF coatings.
    /// Handles spatially varying thermal conductivity and tracks
    /// cumulative melting for porosity closure analysis.
    /// </summary>
    public sealed class FDMCoatingSolver2D : HVOFSim.Domain.Ports.ICoatingSolver
    {
        private readonly double _mu;

        /// <summary>
        /// Create a 2D coating solver.
        /// </summary>
        /// <param name="absorptionCoeff">Optical absorption coefficient mu [1/m].
        /// Controls Beer-Lambert penetration depth. Typical for WC coatings: ~1e4.</param>
        public FDMCoatingSolver2D(double absorptionCoeff = 1e4)
        {
            _mu = absorptionCoeff;
        }

        /// <summary>
        /// Run the 2D laser treatment simulation.
        /// </summary>
        /// <returns>Dictionary-like result with temperature field, melted mask, and duration.</returns>
        public CoatingSolverOutput Solve(
            CoatingGrid coatingGrid,
            CoatingConfig config,
            double laserPower,
            double beamRadius,
            double scanSpeed,
            double absorptivity,
            double coatingRho,
            double coatingCp,
            double tMelt,
            double tAmbient = 25.0)
        {
            var stopwatch = Stopwatch.StartNew();

            int ny = coatingGrid.Ny;
            int nx = coatingGrid.Nx;
            double dx = coatingGrid.Dx;
            double dy = coatingGrid.Dy;

            // --- Build property arrays (spatially varying) ---
            var rho = new double[ny, nx];
            var cp = new double[ny, nx];
            int cr = coatingGrid.CoatingRows;

            for (int y = 0; y < ny; y++)
                for (int x = 0; x < nx; x++)
                {
                    if (y < cr)
                    {
                        rho[y, x] = coatingRho;
                        cp[y, x] = coatingCp;
                    }
                    else
                    {
                        rho[y, x] = config.SubstrateRho;
                        cp[y, x] = config.SubstrateCp;
                    }
                }

            // Copy k map
            var k = new double[ny, nx];
            Array.Copy(coatingGrid.KMap, k, ny * nx);

            // --- CFL Stability ---
            double kappaMax = 0;
            for (int y = 0; y < ny; y++)
                for (int x = 0; x < nx; x++)
                {
                    double kp = k[y, x] / (rho[y, x] * cp[y, x]);
                    if (kp > kappaMax) kappaMax = kp;
                }

            double dtMax = 1.0 / (2.0 * kappaMax * (1.0 / (dx * dx) + 1.0 / (dy * dy)));
            double dt = dtMax * 0.4; // 40% safety margin

            // --- Simulation time ---
            double scanLength = nx * dx;
            double totalTime = scanLength / scanSpeed;
            int nSteps = (int)Math.Ceiling(totalTime / dt);

            int maxSteps = 500_000;
            if (nSteps > maxSteps)
            {
                dt = totalTime / maxSteps;
                nSteps = maxSteps;
            }

            // --- Initialize ---
            var T = new double[ny, nx];
            var melted = new bool[ny, nx];
            for (int y = 0; y < ny; y++)
                for (int x = 0; x < nx; x++)
                    T[y, x] = tAmbient;

            // Coordinate arrays
            var xCoords = new double[nx];
            var yCoords = new double[ny];
            for (int x = 0; x < nx; x++) xCoords[x] = x * dx;
            for (int y = 0; y < ny; y++) yCoords[y] = y * dy;

            // Beer-Lambert depth factor (precomputed)
            var depthFactor = new double[ny];
            for (int y = 0; y < ny; y++)
                depthFactor[y] = Math.Exp(-_mu * yCoords[y]);

            // Gaussian prefactor
            double gaussPrefactor = _mu * 2.0 * absorptivity * laserPower
                                    / (Math.PI * beamRadius * beamRadius);

            // --- Precompute harmonic mean conductivities ---
            int inyH = ny - 2, inxH = nx - 2;
            var kxpDt = new double[inyH, inxH];
            var kxmDt = new double[inyH, inxH];
            var kypDt = new double[inyH, inxH];
            var kymDt = new double[inyH, inxH];
            var rhoCpInv = new double[inyH, inxH];

            for (int y = 0; y < inyH; y++)
                for (int x = 0; x < inxH; x++)
                {
                    int yi = y + 1, xi = x + 1;
                    double kc = k[yi, xi];

                    double kXp = 2.0 * kc * k[yi, xi + 1] / (kc + k[yi, xi + 1] + 1e-30);
                    double kXm = 2.0 * kc * k[yi, xi - 1] / (kc + k[yi, xi - 1] + 1e-30);
                    double kYp = 2.0 * kc * k[yi + 1, xi] / (kc + k[yi + 1, xi] + 1e-30);
                    double kYm = 2.0 * kc * k[yi - 1, xi] / (kc + k[yi - 1, xi] + 1e-30);

                    kxpDt[y, x] = kXp / (dx * dx) * dt;
                    kxmDt[y, x] = kXm / (dx * dx) * dt;
                    kypDt[y, x] = kYp / (dy * dy) * dt;
                    kymDt[y, x] = kYm / (dy * dy) * dt;

                    rhoCpInv[y, x] = 1.0 / (rho[yi, xi] * cp[yi, xi]);
                }

            // --- Time-stepping loop ---
            for (int step = 0; step < nSteps; step++)
            {
                double t = step * dt;
                double laserX = scanSpeed * t;

                // Volumetric heat source for interior
                for (int y = 0; y < inyH; y++)
                {
                    int yi = y + 1;
                    for (int x = 0; x < inxH; x++)
                    {
                        int xi = x + 1;

                        // Gaussian lateral profile
                        double dxL = xCoords[xi] - laserX;
                        double gaussX = Math.Exp(-2.0 * dxL * dxL / (beamRadius * beamRadius));
                        double Q = gaussPrefactor * depthFactor[yi] * gaussX;

                        // FDM diffusion
                        double dTx = kxpDt[y, x] * (T[yi, xi + 1] - T[yi, xi])
                                   - kxmDt[y, x] * (T[yi, xi] - T[yi, xi - 1]);
                        double dTy = kypDt[y, x] * (T[yi + 1, xi] - T[yi, xi])
                                   - kymDt[y, x] * (T[yi, xi] - T[yi - 1, xi]);

                        T[yi, xi] += (dTx + dTy) * rhoCpInv[y, x]
                                   + Q * dt * rhoCpInv[y, x];
                    }
                }

                // Boundary conditions
                for (int x = 0; x < nx; x++)
                {
                    T[0, x] = T[1, x];           // Top: adiabatic
                    T[ny - 1, x] = tAmbient;     // Bottom: fixed
                }
                for (int y = 0; y < ny; y++)
                {
                    T[y, 0] = T[y, 1];           // Left: adiabatic
                    T[y, nx - 1] = T[y, nx - 2]; // Right: adiabatic
                }

                // Cumulative melt tracking
                for (int y = 0; y < ny; y++)
                    for (int x = 0; x < nx; x++)
                        if (T[y, x] > tMelt)
                            melted[y, x] = true;
            }

            stopwatch.Stop();

            return new CoatingSolverOutput(T, melted, stopwatch.Elapsed.TotalSeconds);
        }
    }
}

File: infrastructure/solvers/fdm_coating_solver_2d.py
"""
2D Finite Difference Coating Solver - Stage 2 Laser Treatment.

Specialized 2D FDM solver for laser treatment of HVOF coatings.
Differs from the general-purpose FDMHeatSolver3D in that it:

  1. Operates on a 2D coating grid with spatially varying k (pore-aware)
  2. Uses Beer-Lambert volumetric absorption: Q ~ exp(-mu*y)
  3. Tracks cumulative melt history per cell (melted |= T > T_melt)
  4. Handles coating + substrate as distinct material regions

Governing equation:
    rho*cp * dT/dt = d/dx(k*dT/dx) + d/dy(k*dT/dy) + Q(x,y,t)

Heat source (Gaussian beam scanning at speed v):
    Q(x,y,t) = (2*alpha*P / pi*r^2) * exp(-2*(x - v*t)^2 / r^2) * exp(-mu*y)

Numerical scheme: Forward-Time Central-Space (FTCS) explicit method.

Performance: all static arrays (harmonic-mean k, rho*cp) are precomputed
outside the time loop. Only T, Q, and melt tracking are updated per step.
"""

import sys
import time as time_module

import numpy as np

from domain.coating_entities import CoatingConfig, CoatingGrid


class FDMCoatingSolver2D:
    """
    2D transient heat solver for laser treatment of HVOF coatings.

    Handles spatially varying thermal conductivity and tracks
    cumulative melting for porosity closure analysis.
    """

    def __init__(self, absorption_coeff: float = 1e4):
        """
        Args:
            absorption_coeff: Optical absorption coefficient mu [1/m].
                Controls Beer-Lambert penetration depth.
                Typical for WC coatings: ~1e4 (penetration depth ~100 um).
        """
        self._mu = absorption_coeff

    def solve(
        self,
        coating_grid: CoatingGrid,
        config: CoatingConfig,
        laser_power: float,
        beam_radius: float,
        scan_speed: float,
        absorptivity: float,
        coating_rho: float,
        coating_cp: float,
        t_melt: float,
        t_ambient: float = 25.0,
        verbose: bool = True,
    ) -> dict:
        """
        Run the 2D laser treatment simulation.

        Args:
            coating_grid: Pre-generated coating microstructure.
            config: Coating configuration (substrate properties, etc).
            laser_power: Laser power P [W].
            beam_radius: Beam radius r [m].
            scan_speed: Scan speed v [m/s].
            absorptivity: Surface absorptivity alpha (0-1).
            coating_rho: Coating density rho [kg/m3].
            coating_cp: Coating specific heat cp [J/kg.K].
            t_melt: Melting temperature of the coating matrix [C].
            t_ambient: Ambient/initial temperature [C].
            verbose: Print progress updates.

        Returns:
            dict with keys:
                'temperature': final 2D temperature field
                'melted': cumulative melt mask (bool array)
                'duration_s': wall-clock time
        """
        start = time_module.perf_counter()

        ny, nx = coating_grid.shape
        dx = coating_grid.dx
        dy = coating_grid.dy

        # --- Build property arrays (spatially varying) ---
        rho = np.full((ny, nx), coating_rho, dtype=np.float64)
        cp = np.full((ny, nx), coating_cp, dtype=np.float64)

        # Substrate region has different properties
        cr = coating_grid.coating_rows
        rho[cr:, :] = config.substrate_rho
        cp[cr:, :] = config.substrate_cp

        # Thermal conductivity from the coating grid (already includes pores)
        k = coating_grid.k_map.copy()

        # --- Precompute static arrays (moved out of time loop) ---
        # Harmonic mean conductivities at cell interfaces (STATIC)
        k_xp = 2.0 * k[1:-1, 1:-1] * k[1:-1, 2:] / (k[1:-1, 1:-1] + k[1:-1, 2:] + 1e-30)
        k_xm = 2.0 * k[1:-1, 1:-1] * k[1:-1, :-2] / (k[1:-1, 1:-1] + k[1:-1, :-2] + 1e-30)
        k_yp = 2.0 * k[1:-1, 1:-1] * k[2:, 1:-1] / (k[1:-1, 1:-1] + k[2:, 1:-1] + 1e-30)
        k_ym = 2.0 * k[1:-1, 1:-1] * k[:-2, 1:-1] / (k[1:-1, 1:-1] + k[:-2, 1:-1] + 1e-30)

        # rho*cp for interior (STATIC)
        rho_cp_inv = 1.0 / (rho[1:-1, 1:-1] * cp[1:-1, 1:-1])

        # Precompute diffusion coefficients
        kx_p_dx2 = k_xp / dx**2
        kx_m_dx2 = k_xm / dx**2
        ky_p_dy2 = k_yp / dy**2
        ky_m_dy2 = k_ym / dy**2

        # --- CFL Stability ---
        kappa = k / (rho * cp)
        kappa_max = np.max(kappa)
        dt_max = 1.0 / (2.0 * kappa_max * (1.0 / dx**2 + 1.0 / dy**2))
        dt = dt_max * 0.4  # 40% safety margin

        # --- Simulation time ---
        scan_length = nx * dx
        total_time = scan_length / scan_speed
        n_steps = int(np.ceil(total_time / dt))

        # Cap steps for very long runs
        max_steps = 500_000
        if n_steps > max_steps:
            dt = total_time / max_steps
            n_steps = max_steps
            # Recompute since dt changed
            kx_p_dx2_dt = kx_p_dx2 * dt
            kx_m_dx2_dt = kx_m_dx2 * dt
            ky_p_dy2_dt = ky_p_dy2 * dt
            ky_m_dy2_dt = ky_m_dy2 * dt
        else:
            kx_p_dx2_dt = kx_p_dx2 * dt
            kx_m_dx2_dt = kx_m_dx2 * dt
            ky_p_dy2_dt = ky_p_dy2 * dt
            ky_m_dy2_dt = ky_m_dy2 * dt

        if verbose:
            print(f"    Grid: {ny}x{nx}, dt={dt:.2e}s, "
                  f"n_steps={n_steps}, scan_time={total_time:.4f}s")
            sys.stdout.flush()

        # --- Initialize ---
        T = np.full((ny, nx), t_ambient, dtype=np.float64)
        melted = np.zeros((ny, nx), dtype=bool)

        # Coordinate arrays
        x_coords = np.arange(nx) * dx
        y_coords = np.arange(ny) * dy

        # Beer-Lambert depth factor (precomputed)
        depth_factor = np.exp(-self._mu * y_coords)  # shape (ny,)

        # Gaussian prefactor, including mu for volumetric source [W/m3]
        # Q_vol = mu * (2*alpha*P / pi*r^2) * exp(-2*(x-vt)^2/r^2) * exp(-mu*y)
        gauss_prefactor = self._mu * 2.0 * absorptivity * laser_power / (np.pi * beam_radius**2)

        # Precompute dt * rho_cp_inv for source term
        dt_rho_cp_inv = dt * rho_cp_inv

        # --- Time-stepping loop (optimized) ---
        progress_interval = max(n_steps // 10, 1)

        for step in range(n_steps):
            t = step * dt
            laser_x = scan_speed * t

            # Gaussian lateral profile
            gauss_x = np.exp(-2.0 * (x_coords - laser_x)**2 / beam_radius**2)

            # Volumetric heat source Q(x,y)
            Q_interior = gauss_prefactor * np.outer(
                depth_factor[1:-1], gauss_x[1:-1]
            )

            # FDM diffusion (all precomputed coefficients)
            dT_x = (kx_p_dx2_dt * (T[1:-1, 2:] - T[1:-1, 1:-1])
                     - kx_m_dx2_dt * (T[1:-1, 1:-1] - T[1:-1, :-2]))
            dT_y = (ky_p_dy2_dt * (T[2:, 1:-1] - T[1:-1, 1:-1])
                     - ky_m_dy2_dt * (T[1:-1, 1:-1] - T[:-2, 1:-1]))

            T[1:-1, 1:-1] += (dT_x + dT_y) * rho_cp_inv + Q_interior * dt_rho_cp_inv

            # Boundary conditions (in-place)
            T[0, :] = T[1, :]          # Top: adiabatic
            T[-1, :] = t_ambient       # Bottom: fixed
            T[:, 0] = T[:, 1]          # Left: adiabatic
            T[:, -1] = T[:, -2]        # Right: adiabatic

            # Cumulative melt tracking
            melted |= (T > t_melt)

            # Progress
            if verbose and (step % progress_interval == 0) and step > 0:
                pct = step / n_steps * 100
                peak = np.max(T)
                print(f"      {pct:5.0f}% | step {step}/{n_steps} | "
                      f"peak T = {peak:.0f} C")
                sys.stdout.flush()

        elapsed = time_module.perf_counter() - start

        if verbose:
            print(f"    Done in {elapsed:.1f}s | "
                  f"Peak T = {np.max(T):.0f} C | "
                  f"Melted cells = {np.sum(melted)}")
            sys.stdout.flush()

        return {
            "temperature": T,
            "melted": melted,
            "duration_s": elapsed,
        }
